alf/bin/verify_checkpoint.py

```diff
@@ -0,0 +1,215 @@
+        else:
+            time_step = time_steps[i]
+
+        policy_steps.append(policy_step)
+        policy_state = policy_step.state
+
+    return policy_steps, time_steps
+
+
+def _create_algorithm_and_env(root_dir):
+    """Create algorithm and env from config file."""
+    alf.reset_configs()
+    conf_file = common.get_conf_file()
+    try:
+        common.parse_conf_file(conf_file)
+    except Exception as e:
+        alf.close_env()
+        raise e
+    config = policy_trainer.TrainerConfig(root_dir=root_dir)
+
+    env = alf.get_env()
+    env.reset()
+    data_transformer = create_data_transformer(config.data_transformer_ctor,
+                                               env.observation_spec())
+    config.data_transformer = data_transformer
+
+    # keep compatibility with previous gin based config
+    common.set_global_env(env)
+    observation_spec = data_transformer.transformed_observation_spec
+    common.set_transformed_observation_spec(observation_spec)
+
+    algorithm_ctor = config.algorithm_ctor
+    algorithm = algorithm_ctor(
+        observation_spec=observation_spec,
+        action_spec=env.action_spec(),
+        reward_spec=env.reward_spec(),
+        config=config,
+        env=env)
+    return algorithm, env
+
+
+def main(_):
+    with tempfile.TemporaryDirectory() as root_dir:
+        FLAGS.root_dir = root_dir
+        conf_file = common.get_conf_file()
+        step_num = FLAGS.num_train_iterations
+        ckpt_dir = os.path.join(root_dir, 'ckpt')
+        algorithm1, env1 = _create_algorithm_and_env(root_dir)
+        for i in range(FLAGS.num_train_iterations):
+            # The values of some checkpointed objects (e.g. Normalizer) are
+            # changed by training (i.e. not through random initialization), we
+            # neeed to run the training a few iterations to change those value.
+            # Otherwise they will remain zeros.
+            logging.info("iter=%s" % i)
+            algorithm1.train_iter()
+        ckpt_mngr1 = ckpt_utils.Checkpointer(ckpt_dir, alg=algorithm1)
+        ckpt_mngr1.save(step_num)
+        common.write_config(root_dir)
+
+        FLAGS.gin_file = None
+        FLAGS.conf = None
+        algorithm2, env2 = _create_algorithm_and_env(root_dir)
+        # need one iteration to load checkpoint correctly
+        algorithm2.train_iter()
+        ckpt_mngr2 = ckpt_utils.Checkpointer(ckpt_dir, alg=algorithm2)
+        ckpt_mngr2.load(step_num)
+
+        seed = common.set_random_seed(FLAGS.random_seed)
+        policy_steps1, time_steps = _run_steps(algorithm1, env1,
+                                               FLAGS.num_test_steps)
+        # We calulate policy_steps1 again to make sure policy_steps1 and
+        # policy_steps2 go through exactly same computation sequence so that
+        # they can be compared with each other.
+        common.set_random_seed(seed)
+        policy_steps1, _ = _run_steps(algorithm1, env1, FLAGS.num_test_steps,
```

Comment:
instead of env1 just use None to be clear?

Reply:
> instead of env1 just use None to be clear?

None will break the code because env.batch_size is used in the function

Reply:
Oh, good point, then maybe change env2 below to env1, so it's clear the only difference comes from algorithm1 & 2?